apps/avifdec.c

```diff
@@ -337,78 +333,139 @@ int main(int argc, char * argv[])
         }
-        if (!avifJPEGWrite(outputFilename, decoder->image, jpegQuality, chromaUpsampling)) {
-            goto cleanup;
+
+        if (iccOverrideFilename) {
+            printf("[--icc] Setting ICC profile: %s\n", iccOverrideFilename);
+            result = avifImageSetProfileICC(decoder->image, iccOverride.data, iccOverride.size);
+            if (result != AVIF_RESULT_OK) {
+                fprintf(stderr, "ERROR: Failed to set ICC: %s\n", avifResultToString(result));
+                goto cleanup;
+            }
         }
-    } else if (outputFormat == AVIF_APP_FILE_FORMAT_PNG) {
-        if (!avifPNGWrite(outputFilename, decoder->image, requestedDepth, chromaUpsampling, pngCompressionLevel)) {
+
+        if (decodeAllFrames) {
+            // Create filename for individual frames, in the form path/to/output-000.ext
+            char * lastDot = strrchr(outputFilename, '.');
+            const size_t dotPos = (lastDot != NULL) ? (size_t)(lastDot - outputFilename) : strlen(outputFilename);
+            const char * extension = (lastDot != NULL) ? lastDot + 1 : "";
+            const int maxFilenameWithoutExtensionLength = 1000;
+            const int maxExtensionLength = 10;
+            char frameFilename[1024];
+            int res = snprintf(frameFilename,
+                               sizeof(frameFilename),
+                               "%.*s-%010d.%.*s",
+                               ((int)dotPos > maxFilenameWithoutExtensionLength ? maxFilenameWithoutExtensionLength : (int)dotPos),
+                               outputFilename,
+                               currIndex,
+                               maxExtensionLength,
+                               extension);
+            if (res < 0) {
+                fprintf(stderr, "ERROR: Unable to generate output filename\n");
+                goto cleanup;
+            }
+            if (!avifWriteToFile(outputFormat, frameFilename, decoder->image, rawColor, jpegQuality, pngCompressionLevel, requestedDepth, chromaUpsampling)) {
+                goto cleanup;
+            }
+        } else {
+            if (!avifWriteToFile(outputFormat, outputFilename, decoder->image, rawColor, jpegQuality, pngCompressionLevel, requestedDepth, chromaUpsampling)) {
+                goto cleanup;
+            }
+            if (isSequence && !frameIndexSpecified) {
+                fprintf(stderr,
+                        "INFO: Decoded the first frame of an image sequence with %d frames. To output all frames, use --index all. To silence this message, use --index 0.\n",
+                        decoder->imageCount);
+            }
+            break;
+        }
+        ++currIndex;
+    }
+
+    if (result == AVIF_RESULT_NO_IMAGES_REMAINING) {
+        if (decodeAllFrames) {
+            result = AVIF_RESULT_OK;
+        } else {
+            fprintf(stderr, "ERROR: Frame at index %d requested but the file does not contain enough frames\n", frameIndex);
```

Comment:
`decoder->imageCount` could be mentioned in this error

Reply:
The reason I didn't add it originally was because in my understanding, we don't know if that numer matches the actual number of frames. But you're right that it's useful information in most cases.

Reply:
Looking at the code it seems to be equal to the number of track samples. I have no idea if that is equal to the number of frames in an AVIF image sequence.

Reply:
I think that in theory they should be the same, but the actual movie data might end up having a different number of frames if non compliant? In --info mode we print "(XXX expected frames)" before actually decoding the frames. But this is just a corner case.